controller_manager:
  ros__parameters:
    update_rate: 50  # Hz

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    zebra_zero:
      type: position_controllers/JointGroupPositionController # zebra_zero/RobotController
    
    velocity_controller:
      type: velocity_controllers/JointGroupVelocityController
    
    trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

zebra_zero:
  ros__parameters:
    joints:
      - first_joint
      - second_joint
      - third_joint
      - fourth_joint
      - fifth_joint
      - sixth_joint

    command_interfaces:
      - position

    state_interfaces:
      - position

velocity_controller:
  ros__parameters:
    joints:
      - first_joint
      - second_joint
      - third_joint
      - fourth_joint
      - fifth_joint
      - sixth_joint

    command_interfaces:
      - velocity

    state_interfaces:
      - velocity

trajectory_controller:
  ros__parameters:
    joints:
      - first_joint
      - second_joint
      - third_joint
      - fourth_joint
      - fifth_joint
      - sixth_joint
    
    command_interfaces:
      - velocity

    state_interfaces:
      - position
      - velocity
    
    allow_nonzero_velocity_at_trajectory_end: false

    # error tolerance during motion and at goal
    #constraints:
    #  first_joint: { trajectory: 0.01, goal: 0.005 }
    #  second_joint: { trajectory: 0.01, goal: 0.005 }
    #  third_joint: { trajectory: 0.01, goal: 0.005 }
    #  fourth_joint: { trajectory: 0.01, goal: 0.005 }
    #  fifth_joint: { trajectory: 0.01, goal: 0.005 }
    # sixth_joint: { trajectory: 0.01, goal: 0.005 }

    gains: # Required because we're controlling a velocity interface
      first_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
      second_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
      third_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
      fourth_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
      fifth_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
      sixth_joint: {p: 1.0,  d: 0.0, i: 0.0, ff_velocity_scale: 0.1}
